File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
    static int n;
    static List<Integer>[] graph;
    static boolean[] visited;
    static int[] parent;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        graph = new ArrayList[n+1];
        for(int i=1; i<=n; i++)
            graph[i] = new ArrayList<>();
        visited = new boolean[n+1];
        parent = new int[n+1];
        for(int i=1; i<n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            graph[x].add(y);
            graph[y].add(x);
        }

        dfs(1);

        for(int i=2; i<=n; i++)
            System.out.println(parent[i]);
    }

    static void dfs(int v) {
        visited[v] = true;
        for(int i : graph[v]) {
            if(!visited[i]) {
                parent[i] = v;
                dfs(i);
            }
        }
    }
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.kt
fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val graph = Array(n + 1) { ArrayList<Int>() }
    repeat(n - 1) {
        val (u, v) = readLine().split(" ").map { it.toInt() }
        graph[u].add(v)
        graph[v].add(u)
    }
    
    val parents = IntArray(n + 1)
    val visited = BooleanArray(n + 1)
    
    val queue = ArrayDeque<Int>()
    queue.add(1)
    visited[1] = true
    
    while (queue.isNotEmpty()) {
        val current = queue.removeFirst()
        
        for (neighbor in graph[current]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true
                parents[neighbor] = current
                queue.add(neighbor)
            }
        }
    }
    
    val sb = StringBuilder()
    for (i in 2..n) {
        sb.appendLine(parents[i])
    }
    print(sb)
}